SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSMessageReceiver.h"
+#import "OWSSignalServiceProtos.pb.h"
+#import "TSDatabaseView.h"
+#import "TSMessagesManager.h"
+#import "TSStorageManager.h"
+#import "TSYapDatabaseObject.h"
+#import <YapDatabase/YapDatabaseConnection.h>
+#import <YapDatabase/YapDatabaseTransaction.h>
+#import <YapDatabase/YapDatabaseView.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Persisted data model
+@class OWSSignalServiceProtosEnvelope;
+
+@interface OWSMessageProcessingJob : TSYapDatabaseObject
+
+@property (nonatomic, readonly) OWSSignalServiceProtosEnvelope *envelopeProto;
+@property (nonatomic, readonly) NSDate *createdAt;
```

Comment:
I thought about this for a while, and it's not clear to me what's best.

You bring up so good points, but my conclusions was that even if it means the resultant received history is "out of order" with respect to when it was sent, we never want to insert a new message "up" in the conversation, because the recipient might miss the message.

In practice, we currently process the messages in the order in which they were received due to the YapDBView which sorts by createdAt. So to be clear, currently we're processing messages in the same order before and after this PR.

I'm inclined to leave it as is unless we can make a change this is (at least mostly) uncontroversial.

And specifically in the case of the group info update, we probably want the ordering to reflect *processing* time, so the "group update" would appear, *and then* the message. Right?

I think one thing at the root of this is our distinction between sort order and display date. To get in line with android, we short sort by received date, but display the sending users timestamp in the UI. There is a problem when the senders clock is busted, but in practice I guess this seems to have provided more value than always showing the "Received at" time. 

I think a similar approach (sorting by received at, but displaying senders time) would better serve our purposes in the long run - but it probably only makes sense once we have a visible per message timestamp. Otherwise with only the current date headers, it would be easy to have an **even more** confusing situation when the sender's clock is off e.g. in a group.